src/backends/plonky2/circuits/mainpod.rs

```diff
@@ -61,39 +61,112 @@ struct OperationVerifyGadget {
     params: Params,
 }
 
-impl OperationVerifyGadget {
-    /// Checks whether the first `N` arguments to an op are ValueOf
-    /// statements, returning a boolean target indicating whether this
-    /// is the case as well as the value targets derived from each
-    /// argument.
-    fn first_n_args_as_values<const N: usize>(
-        &self,
+const MAX_VALUE_ARGS: usize = 3;
+
+struct StatementArgCache {
+    rhs: ValueTarget,
+    lhs: StatementArgTarget,
+    valid: BoolTarget,
+}
+
+struct StatementCache {
+    equations: [StatementArgCache; MAX_VALUE_ARGS],
+    first_n_equations_valid: [BoolTarget; MAX_VALUE_ARGS],
+    op_args: Vec<StatementTarget>,
+}
+
+impl StatementCache {
+    fn new(
+        params: &Params,
         builder: &mut CircuitBuilder,
-        resolved_op_args: &[StatementTarget],
-    ) -> (BoolTarget, [ValueTarget; N]) {
-        let arg_is_valueof = resolved_op_args[..N]
-            .iter()
-            .map(|arg| {
-                let st_type_ok =
-                    arg.has_native_type(builder, &self.params, NativePredicate::ValueOf);
-                let value_arg_ok = builder.statement_arg_is_value(&arg.args[1]);
-                builder.and(st_type_ok, value_arg_ok)
-            })
-            .collect::<Vec<_>>();
-        let first_n_args_are_valueofs = arg_is_valueof
-            .into_iter()
-            .reduce(|a, b| builder.and(a, b))
-            .expect("No args specified.");
-        let values = array::from_fn(|i| resolved_op_args[i].args[1].as_value());
-        (first_n_args_are_valueofs, values)
+        op: &OperationTarget,
+        st: &StatementTarget,
+        prev_statements: &[StatementTarget],
+    ) -> Self {
+        let op_args = if prev_statements.is_empty() {
+            (0..params.max_operation_args)
+                .map(|_| StatementTarget::new_native(builder, params, NativePredicate::None, &[]))
+                .collect_vec()
+        } else {
+            // `op.args` is a vector of arrays of length 1, so `.flatten()` is just
+            // converting a length 1 array into a scalar.
+            op.args
+                .iter()
+                .flatten()
+                .map(|&i| builder.vec_ref(params, prev_statements, i))
+                .collect::<Vec<_>>()
+        };
+        let equations = array::from_fn(|i| {
+            if i < params.max_operation_args && i < params.max_statement_args {
```

Comment:
The `equations` array has `len=3`.
We have NativePredicates like `SumOf` that take 3 arguments.
And we have NativeOperations like `SumOf` that take 3 arguments.
To simplify things (and avoid potential edge-cases) I would like to assume that `params.max_operation_args >= 3` and `params.max_statement_args >= 3`.

Currently we don't validate the `Params` when they are created, but some methods do some checks on the params.

In this case I think the code and logic would be simplified if we just do:
```rust
assert!(params.max_operation_args <= 3);
assert!(params.max_statement_args <= 3);
```
and then skip this if/else.